packages/shared/src/features/onboarding/hooks/useFunnelTracking.ts

```diff
@@ -0,0 +1,172 @@
+export type TrackOnNavigate = (event: {
+  from: FunnelStep['id'];
+  to: FunnelStep['id'];
+  timeDuration: number;
+  type: FunnelStepTransitionType;
+}) => void;
+
+interface USeFunnelTrackingProps {
+  funnel: FunnelJSON;
+}
+
+interface UseFunnelTrackingReturn {
+  trackOnClickCapture: TrackOnMouseCapture;
+  trackOnHoverCapture: TrackOnMouseCapture;
+  trackFunnelEvent: TrackOnEvent;
+  trackOnNavigate: TrackOnNavigate;
+  trackOnScroll: TrackOnScroll;
+}
+
+const trackOnMouseCapture = ({
+  selector,
+  eventName,
+  trackFunnelEvent,
+}: {
+  selector: string;
+  eventName:
+    | FunnelEventName.ClickFunnelElement
+    | FunnelEventName.HoverFunnelElement;
+  trackFunnelEvent: TrackOnEvent;
+}): TrackOnMouseCapture => {
+  return (event) => {
+    if (!(event.target instanceof HTMLElement)) {
+      return;
+    }
+
+    const trackedElement = event.target.closest(selector);
+
+    if (!trackedElement) {
+      return;
+    }
+
+    trackFunnelEvent({
+      name: eventName,
+      details: {
+        target_type:
+          trackedElement.tagName.toLowerCase() as keyof HTMLElementTagNameMap,
+        ...(trackedElement.id ? { target_id: trackedElement.id } : {}),
+      },
+    });
+  };
+};
+
+export const useFunnelTracking = ({
+  funnel,
+}: USeFunnelTrackingProps): UseFunnelTrackingReturn => {
+  const { logEvent } = useLogContext();
+  const position = useAtomValue(funnelPositionAtom);
+  const step = useMemo(
+    () => getFunnelStepByPosition(funnel, position),
+    [funnel, position],
+  );
+
+  const trackFunnelEvent: TrackOnEvent = useCallback(
+    (event) => {
+      const commonTrackingProps = {
+        funnel_id: funnel.id,
+        funnel_version: funnel.version,
+        session_id: '', // todo: implement sessionId
+        step_id: step?.id,
+        step_type: step?.type,
+      };
+
+      logEvent({
+        event_name: event.name,
+        extra: JSON.stringify({
+          ...commonTrackingProps,
+          ...('details' in event ? event.details : {}),
+        }),
+      });
+    },
+    [funnel, logEvent, step],
+  );
+
+  const trackOnClickCapture: TrackOnMouseCapture = trackOnMouseCapture({
+    selector: '[data-track-click]',
+    eventName: FunnelEventName.ClickFunnelElement,
+    trackFunnelEvent,
+  });
+
+  const trackOnHoverCapture: TrackOnMouseCapture = trackOnMouseCapture({
+    selector: '[data-track-hover]',
+    eventName: FunnelEventName.HoverFunnelElement,
+    trackFunnelEvent,
+  });
+
+  const trackOnScroll: TrackOnScroll = useCallback(() => {
+    trackFunnelEvent({
+      name: FunnelEventName.ScrollFunnel,
+      details: {
+        scroll_y: globalThis.scrollY,
+      },
+    });
+  }, [trackFunnelEvent]);
+
+  const trackOnNavigate: TrackOnNavigate = useCallback(
+    (event) => {
+      trackFunnelEvent({
+        name: FunnelEventName.TransitionFunnel,
+        details: {
+          target_type: event.type,
+          target_id: event.to,
+          event_duration: event.timeDuration,
```

Comment:
This was supposed to be outside of extra, but i'll update the excel
```suggestion
          duration: event.timeDuration,
```

Reply:
Done ✔️